### ISSUE TITLE
Using CUB package (instead of vendoring)

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/cupy-feedstock#60]
Recently [`cub`]( https://github.com/conda-forge/cub-feedstock ) got added to `conda-forge`. It would make sense (once we start building with `cub` again) to use the `cub` package instead of [downloading our own copy]( https://github.com/conda-forge/cupy-feedstock/blob/20e477abcdd677c1748acc1c3db408b7ea49e032/recipe/meta.yaml#L16-L18 ).

**Comment by leofang** [comment]
I am not sure why we wanna use it. For CUDA < 11, if CUB is bundled (cupy/cupy#2584) then we just use the bundled version. For CUDA 11+, CUB is version-locked to Thrust, and Bryce has recommended to only use the version that comes with Thrust from the CUDA Toolkit, which I think would be available in the docker image.

**Comment by jakirkham** [comment]
We could use selectors to grab it when appropriate and skip otherwise.

**Comment by leofang** [comment]
By "when appropriate" you meant depending on CuPy and CUDA versions?

**Comment by jakirkham** [comment]
Yeah that's right.

**Comment by leofang** [comment]
By the way, with https://github.com/nsls-ii-forge/cupy-feedstock I got a lot more confident of brining back CUB. It's enabled there and full tests are run. Only a few minor, irrelevant issues occurred.

**Comment by leofang** [comment]
Opened an issue for the `cub` package: https://github.com/conda-forge/cub-feedstock/issues/4.

**Comment by jakirkham** [comment]
Great! Glad to hear CUB support is working. 😄

I think we really need to get your PR ( https://github.com/cupy/cupy/pull/2584 ) across the line. Happy to help however I can.

Ok, thanks will take a look 🙂

**Comment by leofang** [comment]
> I think we really need to get your PR ( [cupy/cupy#2584](https://github.com/cupy/cupy/pull/2584) ) across the line. 

Yeah I agree. We probably discussed somewhere else that https://github.com/chainer/chainer-test/pull/564 seems to be the only blocker.

Once it's merged, we can start discussing how to address the Thrust-CUB locking issue. For example, we could order the search path like this (which is currently done): 
```
-I/usr/local/cuda/include -Icupy/core/include/cupy
``` 
so that for CUDA 11+ we use those from the CUDA Toolkit, and for CUDA <11 we use Thrust from Toolkit + bundled CUB (which is considered interoperable). We could also just ignore the bundled CUB for CUDA 11+. That discussion is less urgent I think.

**Comment by jakirkham** [comment]
Is that still a concern given [PfN is moving away from Chainer to PyTorch]( https://chainer.org/announcement/2019/12/05/released-v7.html )?

**Comment by jakirkham** [comment]
Yeah I think those packages can become empty packages for CUDA 11+. So it will make it easy for us to depend on them for different CUDA builds.

**Comment by leofang** [comment]
> Is that still a concern given [PfN is moving away from Chainer to PyTorch](https://chainer.org/announcement/2019/12/05/released-v7.html)?

Right, because Chainer-test sets up the CI for both Chainer and CuPy.

**Comment by leofang** [comment]
Hi @jakirkham The discussion in cupy/cupy#3445 concludes that CuPy will no longer allow user-supplied CUB location. The official policy is to use the bundled CUB -- either with CuPy or with CUDA (11+). I think it makes sense because then those combinations will be tested by CuPy's CI, whereas conda's CUB won't be tested anywhere.

Given that, I'm gonna close this issue. Please feel free to jump in there if you find we overlooked anything in the discussion. Then, we can consider reopening this one. Thanks!
